src/named_pipe.rs

```diff
@@ -0,0 +1,208 @@
+        let component_input_fifo_path = self.dir.path().join(component_input_fifo_filename);
+
+        let component_output_fifo_filename = format!("OUT_{}", pipe_num);
+        let component_output_fifo_path = self.dir.path().join(component_output_fifo_filename);
+
+        mkfifo(
+            &component_input_fifo_path,
+            Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+        )?;
+        mkfifo(
+            &component_output_fifo_path,
+            Mode::S_IRWXU | Mode::S_IRWXG | Mode::S_IRWXO,
+        )?;
+
+        debug!(
+            "Creating new pipe I = {:?}, O = {:?}",
+            component_input_fifo_path, component_output_fifo_path
+        );
+
+        Ok(NamedPipe {
+            component_input_fifo_path,
+            component_output_fifo_path,
+
+            component_input_fifo_file: None,
+            component_output_fifo_file: None,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct NamedPipe {
+    component_input_fifo_path: PathBuf,
+    component_output_fifo_path: PathBuf,
+
+    component_input_fifo_file: Option<File>,
+    component_output_fifo_file: Option<File>,
+}
+
+const PIPE_CREATION_TIMEOUT_MS: u64 = 100;
+const PIPE_IO_TIMEOUT_MS: u64 = 1000;
+const PIPE_POLL_INTERVAL_MS: u64 = 2;
+
+const BUF_SIZE: usize = 256;
+
+impl NamedPipe {
+    fn get_fds(&mut self) -> Result<(RawFd, RawFd), WorkerError> {
+        let deadline = Instant::now() + Duration::from_millis(PIPE_CREATION_TIMEOUT_MS);
+
+        while self.component_output_fifo_file.is_none() && Instant::now() < deadline {
+            let c_out_res = OpenOptions::new()
+                .read(true)
+                .custom_flags(OFlag::O_NONBLOCK.bits())
+                .open(&self.component_output_fifo_path);
+
+            debug!("Opening component output {:?}", c_out_res);
+
+            self.component_output_fifo_file = c_out_res.ok();
+
+            sleep(Duration::from_millis(PIPE_POLL_INTERVAL_MS))
+        }
+
+        while self.component_input_fifo_file.is_none() && Instant::now() < deadline {
+            let c_in_res = OpenOptions::new()
+                .write(true)
+                .custom_flags(OFlag::O_NONBLOCK.bits())
+                .open(&self.component_input_fifo_path);
+
+            debug!("Opening component input {:?}", c_in_res);
+
+            self.component_input_fifo_file = c_in_res.ok();
+
+            sleep(Duration::from_millis(PIPE_POLL_INTERVAL_MS))
+        }
+
+        if let (Some(c_in), Some(c_out)) =
+            (&self.component_input_fifo_file, &self.component_output_fifo_file)
+        {
+            Ok((c_in.as_raw_fd(), c_out.as_raw_fd()))
+        } else {
+            Err(WorkerErrorKind::OperationTimedOut.into())
+        }
+    }
+
+    pub fn write(&mut self, v: &[u8]) -> Result<(), WorkerError> {
+        let (c_in_fd, _) = self.get_fds()?;
+
+        let deadline = Instant::now() + Duration::from_millis(PIPE_IO_TIMEOUT_MS);
+
+        assert_eq!(v.last(), Some(&b'\n'));
```

Comment:
Perhaps this layer should handle terminating with a newline? Need to think about the preconditions/postconditions of this API